<a href="https://colab.research.google.com/github/updalla-apshir/machine-learning-projects/blob/main/4_7_Handling_Imbalanced_Dataset_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Imbalance Dataset:
a dataset with an unequal class distribution

In [ ]:
# importing the dependencies
import numpy as np
import pandas as pd

In [ ]:
# loading a dataset to a pandas DataFrame
creditCard_data = pd.read_csv('credit_data.csv')


In [ ]:
creditCard_data.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0.0
1,0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0.0
2,1,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0.0
3,1,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0.0
4,2,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0.0


In [ ]:
creditCard_data['Class'].value_counts()

,count
Class,
0.0,118675
1.0,245


This is Highly Imbalanced Dataset

0 -> Legit Transactions

1 -> Fraudulent Transactions

In [ ]:
legit = creditCard_data[creditCard_data.Class == 0]
fraud = creditCard_data[creditCard_data.Class == 1]

In [ ]:
print(legit.shape)
print(fraud.shape)


(118675, 31)
(245, 31)


Under-sampling

Build a simple dataset containing similar distribution of legit & fraudulent Transactions

number of fraudulent Transaction -> 245

In [ ]:
legit_sample = legit.sample(250)


In [ ]:
legit_sample

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
560,417,-0.791993,1.054007,-0.110012,-1.234186,2.751355,3.443064,0.055313,-0.105660,-0.306074,...,0.425088,-1.178829,-0.032154,0.928394,-0.128566,0.046299,0.069745,-0.037621,0.89,0.0
111357,72200,1.120607,0.138940,0.963673,1.943452,-0.566023,-0.097860,-0.241817,0.122793,-0.065271,...,-0.283436,-0.728337,0.203663,0.413840,0.031881,0.763103,-0.050699,0.011308,9.16,0.0
40174,40125,-0.067417,0.602581,0.682480,1.375690,0.546689,1.306665,0.023801,0.501602,-0.057021,...,0.180261,0.980493,0.297453,-0.665712,-1.595575,0.530057,0.353033,0.271525,29.56,0.0
114911,73669,-1.694731,0.354838,0.149354,-0.198226,-2.627326,0.995440,1.520382,0.769352,-0.421687,...,0.255466,0.574613,0.432320,0.268693,-0.771273,0.199713,-0.003612,-0.120985,493.31,0.0
39358,39803,-2.599569,2.800236,-1.026366,-0.333287,-0.171991,2.004893,-3.564713,-6.777160,-1.814288,...,-3.539164,1.560874,0.476009,-1.361128,-0.417098,0.326921,-0.387388,0.106259,1.50,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
64051,50999,0.919068,-0.460092,1.706972,1.270324,-0.732696,1.950031,-1.349882,0.842849,1.336144,...,0.306866,1.295213,0.066004,-0.607221,0.092228,-0.079328,0.166947,0.029476,6.15,0.0
40106,40096,-0.726066,-0.608919,2.016353,-2.606823,-0.642338,0.387095,-0.777382,0.324828,-2.390989,...,0.143972,0.768363,-0.051087,-0.291391,-0.190160,-0.225172,0.018188,0.107788,5.20,0.0
81354,58901,-0.761954,1.204124,0.759431,-0.127442,0.038526,-0.051850,0.164330,0.689826,-0.722279,...,-0.163569,-0.581089,0.050174,-0.377266,-0.210109,0.108473,0.124977,0.018020,14.99,0.0
80534,58530,-1.590994,1.589185,-0.445582,0.846818,-0.746814,0.138052,-0.275106,1.320761,-1.058751,...,0.262961,0.570860,0.122763,-0.286842,-0.072561,-0.292334,-0.197252,-0.235165,67.00,0.0


Concatenate the Two Dataframes

In [ ]:
new_data = pd.concat([legit_sample,fraud], axis=0)

In [ ]:
new_data

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
560,417,-0.791993,1.054007,-0.110012,-1.234186,2.751355,3.443064,0.055313,-0.105660,-0.306074,...,0.425088,-1.178829,-0.032154,0.928394,-0.128566,0.046299,0.069745,-0.037621,0.89,0.0
111357,72200,1.120607,0.138940,0.963673,1.943452,-0.566023,-0.097860,-0.241817,0.122793,-0.065271,...,-0.283436,-0.728337,0.203663,0.413840,0.031881,0.763103,-0.050699,0.011308,9.16,0.0
40174,40125,-0.067417,0.602581,0.682480,1.375690,0.546689,1.306665,0.023801,0.501602,-0.057021,...,0.180261,0.980493,0.297453,-0.665712,-1.595575,0.530057,0.353033,0.271525,29.56,0.0
114911,73669,-1.694731,0.354838,0.149354,-0.198226,-2.627326,0.995440,1.520382,0.769352,-0.421687,...,0.255466,0.574613,0.432320,0.268693,-0.771273,0.199713,-0.003612,-0.120985,493.31,0.0
39358,39803,-2.599569,2.800236,-1.026366,-0.333287,-0.171991,2.004893,-3.564713,-6.777160,-1.814288,...,-3.539164,1.560874,0.476009,-1.361128,-0.417098,0.326921,-0.387388,0.106259,1.50,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
112840,72824,-1.111495,-0.257575,2.250210,1.152671,0.432904,1.254126,-0.584163,-0.609682,1.014602,...,0.862913,0.927825,-0.343058,-0.256268,-0.600742,-0.180331,0.026762,-0.358335,45.03,1.0
114271,73408,-2.869795,1.335667,-1.009530,1.693885,-0.741480,-0.796773,-2.614242,1.066636,-1.135497,...,0.528421,0.228027,-0.749298,-0.067179,0.215792,-0.453651,0.349191,0.198862,16.48,1.0
116139,74159,-1.548788,1.808698,-0.953509,2.213085,-2.015728,-0.913457,-2.356013,1.197169,-1.678374,...,0.855138,0.774745,0.059037,0.343200,-0.468938,-0.278338,0.625922,0.395573,76.94,1.0
116404,74262,-2.250535,2.365755,-2.955491,0.089791,-2.830745,-0.844462,-0.174062,-0.407138,0.174216,...,0.256560,-0.466245,0.291105,0.242567,-1.279094,-1.123534,-0.630977,0.326839,311.28,1.0


In [ ]:
new_data['Class'].value_counts()

,count
Class,
0.0,250
1.0,245
